man/man5/crontab.5
.TH CRONTAB 5
.SH NAME
crontab \- table of jobs to be performed by cron
.SH SYNOPSIS
.nf
.ft B
/usr/lib/crontab
/usr/local/lib/crontab
/var/lib/crontab
/var/opt/\fIname\fP/lib/crontab\ \ \fR(Minix-vmd only)\fB
/usr/spool/crontabs/\fIuser\fP
.ft R
.fi
.SH DESCRIPTION
The
.BR cron (8)
daemon runs jobs at regular intervals.  These jobs are listed in
.B crontab
files.  The format of entries in a crontab file are five fields of numbers
specifying the minute (0\-59), hour (0\-23), day of the month (1\-31), month
(1\-12), and day of the week (0\-6 with 0 = Sunday) that a task must be
executed.  The task to be executed follows as a shell command.
.PP
The time numbers can be given as a comma separated list of simple numbers,
ranges ("2\-5" is the same as "2,3,4,5"), and repeats ("2:5" means
"2,7,12,17,22" in the hour field).  A repeat is cyclic affair, i.e. 2:5
and 12:5 are the same thing.  A single "*" can be used in a field to
indicate all valid numbers in that field, so it translates to "always".  In
the minute field you can use "?" for the current minute that the crontab
file is loaded.  It can be used in a repeat, i.e. "?:10" for every 10
minutes.  This keeps machines with identical crontabs from executing tasks
at exactly the same time, causing a burst of traffic if anything is done
over a network.
.PP
If a given time is valid in all five fields then a command is executed.
Here are a few examples that illustrate the possibilities:
.PP
.if t .RS
.if t .ft C
.nf
# min hour mday mon wday command
   ?   3    *    *   *   /usr/etc/daily  # Daily system cleanup
   0   *    *    *   *   date            # Print date on the hour
  30   4    *    *  2\-6  /var/etc/backup # After workdays on 4:30
   0   9   25   12   *   \-u ast sing     # Andy sings on Xmas morning
   0   0   13    *   5   echo Beware!    # For the superstitious
.fi
.if t .ft P
.if t .RE
.PP
The command may optionally be prefixed by
.BI \-u " user"
to specify under which user the command should be run.  Commands from
crontabs in the spool directory are always run under the id of the crontab's
owner, the
.B \-u
flag is ignored.
.PP
A command can be placed on the same line as the time fields, or on the next
line indented by one TAB character.  (A TAB, not eight spaces.)  More TAB
indented lines can be added for a multiline command.  The tabs are removed
from the command when passed to the shell.  If a command is put on the same
line as the time fields then percent characters are changed into newlines,
this is not done for a TAB indented command.  The following three entries
give the same output:
.PP
.RS
.if t .ft C
.nf
.ta +8n
0 12  *  *  *   echo 'Hello'; echo '  World!'
#1
0 12  *  *  *   echo 'Hello%  World!'  #2
0 12  *  *  *        #3
	cat <<EOF    #4
	Hello
	\&  World!
	EOF
.fi
.if t .ft P
.RE
.PP
Comments start with a "#" character and continue until end of line.  They,
excess whitespace, and empty lines are ignored.  Of the comments in the
example above #1 and #3 are ignored by
.BR cron ,
but #2 and #4 are not recognized as comments, but are seen as part of a
command and are passed to the shell who then happens to ignore them.  There
is no interpretation of command characters other than the percent in a
oneliner.  The time fields must all be on the same line.
.SH FILES
.TP 25n
.B /usr/lib/crontab
Main Minix crontab file.
.TP
.B /usr/local/lib/crontab
Local jobs for all systems in an organization.
.TP
.B /var/lib/crontab
System specific jobs.
.TP
.B /var/opt/\fIname\fP/lib/crontab
Per package jobs for Minix-vmd.
.TP
.B /usr/lib/packages
List of installed packages.
.TP
.B /usr/spool/crontabs/\fIuser\fP
Per user jobs.
.SH "SEE ALSO"
.BR crontab (1),
.BR cron (8).
.SH NOTES
The "?" in the minute field, the repeat field (e.g. "2:5"), TAB indented
multiline commands and the
.B \-u
option are unique to this cron implementation.  This doesn't mean you
shouldn't use these features, but just that you should be aware of the
differences with other systems.  You are even advised to use these features
and avoid the percent hack for multiline commands.
.PP
Other crons allow one to specify input to a job in some way, something this
cron can't.  Simply use the << shell feature to do that.  Other crons often
choke on empty lines.
.PP
It is a common bug to use 0 for Sunday instead of 7.  This cron, like most
other crons out there accepts this without comment.
.PP
A job is not reissued until a previous instance of it has exited.  The next
time to execute is computed from the previous time it ran.  If job issuing
lags behind on the system time then the next time to run it is computed from
the current system time.
.SH AUTHOR
Kees J. Bot (kjb@cs.vu.nl)

.\"
.\" $PchId: crontab.5,v 1.3 2000/07/17 18:53:05 philip Exp $


commands/cron/crontab.c
/*	crontab 1.2 - user crontab manipulation		Author: Kees J. Bot
 *								12 Jan 1997
 */
#define nil ((void*)0)
#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <pwd.h>
#include <sys/stat.h>
#include "misc.h"
#include "tab.h"

#if __minix && !__minix_vmd
#define seteuid(uid)	((uid),0)	/* Minix can't fiddle with uids. */
#endif

static int opentab(int uid, char *file, int how)
/* Open a crontab file under the given uid.  How is 'r' or 'w'.  Return
 * the result of open(2).
 */
{
	uid_t safe_uid;
	int flags, r, err;

	switch (how) {
	case 'r':	flags= O_RDONLY;			break;
	case 'w':	flags= O_WRONLY | O_CREAT | O_TRUNC;	break;
	default:	errno= EINVAL;				return -1;
	}

#if __minix && !__minix_vmd
	/* Standard Minix has no saved uid, so use the lousy old access(). */
	if (uid != 0) {
		if (access(file, how == 'r' ? R_OK : W_OK) < 0) return -1;
	}
#endif

	safe_uid= geteuid();
	seteuid(uid);
	r= open(file, flags, 0666);
	err= errno;
	seteuid(safe_uid);
	errno= err;
	return r;
}

static void copytab(int fd_in, char *file_in, int fd_out, char *file_out)
/* Copy one open file to another.  Complain and exit on errors. */
{
	ssize_t r, w;
	char buf[1024];

	while ((r= read(fd_in, buf, sizeof(buf))) > 0) {
		w= 0;
		while (w < r) {
			if ((r= write(fd_out, buf+w, r-w)) <= 0) {
				fprintf(stderr,
				"%s: Write error on %s: %s\n",
					prog_name,
					file_out,
					r == 0 ? "End of file"
							: strerror(errno));
				exit(1);
			}
			w+= r;
		}
	}
	if (r < 0) {
		fprintf(stderr, "%s: Read error on %s: %s\n",
			prog_name, file_in, strerror(errno));
		exit(1);
	}
}

static void usage(void)
{
	fprintf(stderr,
		"Usage: %s -c [user] file  # Change crontab\n"
		"       %s -l [user]       # List crontab\n"
		"       %s -r [user]       # Remove crontab\n"
		"       %s -p              # Tell cron to reload\n",
		prog_name, prog_name, prog_name, prog_name);
	exit(1);
}

int main(int argc, char **argv)
{
	int i;
	int cflag, lflag, rflag, pflag;
	uid_t uid;
	char *user, *file;
	struct passwd *pw;
	static char SPOOLDIR[]= "/usr/spool/crontabs";
	char tabfile[sizeof(SPOOLDIR) + NAME_MAX];

	prog_name= strrchr(argv[0], '/');
	if (prog_name == nil) prog_name= argv[0]; else prog_name++;

	cflag= lflag= rflag= pflag= 0;
	i= 1;
	while (i < argc && argv[i][0] == '-') {
		char *opt= argv[i++] + 1;

		if (opt[0] == '-' && opt[1] == 0) break;	/* -- */

		while (*opt != 0) switch (*opt++) {
		case 'c':	cflag= 1;	break;
		case 'l':	lflag= 1;	break;
		case 'r':	rflag= 1;	break;
		case 'p':	pflag= 1;	break;
		default:	usage();
		}
	}
	if (cflag + lflag + rflag + pflag != 1) usage();

	user= file= nil;
	if (!pflag && i < argc) user= argv[i++];
	if (cflag) {
		if (user == nil) usage();
		if (i < argc) {
			file= argv[i++];
		} else {
			file= user;
			user= nil;
		}
	}
	if (i != argc) usage();

	if (geteuid() != 0) {
		fprintf(stderr, "%s: No root privileges?\n", prog_name);
	}
	uid= getuid();
	if (user == nil) {
		if ((pw= getpwuid(uid)) == nil) {
			fprintf(stderr,
				"%s: Don't know who you (uid %lu) are!\n",
				prog_name, (unsigned long) uid);
			exit(1);
		}
	} else {
		if ((pw= getpwnam(user)) == nil) {
			fprintf(stderr,
				"%s: Don't know who you (%s) are!\n",
				prog_name, user);
			exit(1);
		}
	}
	if (uid != 0 && pw->pw_uid != uid) {
		fprintf(stderr,
		"%s: Only root can change the crontabs of others!\n",
			prog_name);
		exit(1);
	}
	user= pw->pw_name;
	uid= pw->pw_uid;
	seteuid(uid);
	umask(0077);

	selectlog(STDERR);
	sprintf(tabfile, "%s/%s", SPOOLDIR, user);

	if (lflag) {
		int fd;

		if ((fd= opentab(0, tabfile, 'r')) < 0) {
			fprintf(stderr, "%s: Can't open %s: %s\n",
				prog_name, tabfile, strerror(errno));
			exit(1);
		}
		copytab(fd, tabfile, 1, "stdout");
		close(fd);
	}

	if (rflag) {
		seteuid(0);
		if (unlink(tabfile) < 0) {
			fprintf(stderr, "%s: Can't remove %s: %s\n",
				prog_name, tabfile, strerror(errno));
			exit(1);
		}
		seteuid(uid);
		printf("Crontab of %s removed\n", user);
		pflag= 1;
	}

	if (cflag) {
		int fd1, fd2;

		if ((fd1= opentab(uid, file, 'r')) < 0) {
			fprintf(stderr, "%s: Can't open %s: %s\n",
				prog_name, file, strerror(errno));
			exit(1);
		}

		/* Try to parse the new crontab file.  If the parsing
		 * succeeds then 'crontabs' will be non-null.
		 */
		tab_parse(file, user);
		tab_purge();
		if (crontabs == nil) exit(1);

		if ((fd2= opentab(0, tabfile, 'w')) < 0) {
			fprintf(stderr, "%s: Can't open %s: %s\n",
				prog_name, tabfile, strerror(errno));
			exit(1);
		}
		copytab(fd1, file, fd2, tabfile);
		close(fd1);
		close(fd2);
		printf("New crontab for %s installed\n", user);
		pflag= 1;
	}

	if (pflag) {
		/* Alert cron to the new situation. */
		FILE *fp;

		seteuid(0);
		if ((fp= fopen(PIDFILE, "r")) != NULL) {
			unsigned long pid;
			int c;

			pid= 0;
			while ((c= fgetc(fp)) != EOF && c != '\n') {
				if ((unsigned) (c - '0') >= 10) {
					pid= 0; break;
				}
				pid= 10*pid + (c - '0');
				if (pid >= 30000) { pid= 0; break; }
			}
			if (pid > 1 && kill((pid_t) pid, SIGHUP) == 0) {
				pflag= 0;
			}
		}
		seteuid(uid);
		if (pflag) {
			fprintf(stderr,
			"%s: Alerting cron has failed; cron still running?\n",
				prog_name);
			exit(1);
		}
		printf("Cron signalled to reload tables\n");
	}
	return 0;
}

/*
 * $PchId: crontab.c,v 1.4 2000/07/17 18:54:50 philip Exp $
 */
